### PR TITLE
APEXCORE-626 shutdown-app in cli supports app-name as argument

### PR DESCRIPTION
**Pull request by florianschmidt1994**
The command shutdown-app now supports not only the app-id but also the
app name as an argument.

This commit adds both the functionality as well as the documentation to
the cli help messages

### CODE REVIEW

**Comment by pramodin** [comment]
Are there any unit tests for shutdown?

**Comment by florianschmidt1994** [comment]
I looked into `ApexCliTest` and `ApexCliMiscTest` and they both don't seem to contain tests related to the shutdown command

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (tushargosavi) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: app = getApplication(appNameOrId)
if (app == null) {
     app = getApplicationByName(appNameOrId)
}

To avoid calling getApplication and getApplicationByName twice.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (tushargosavi) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:2187]: this function could be used in KillAppCommand#execute too

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (tushargosavi) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:743]: Similar change is needed at
     globalCommands.put("shutdown-app", new CommandSpec(new ShutdownAppCommand(),
         new Arg[]{new Arg("app-id")},
         new Arg[]{new VarArg("app-id")},

line number 634

**Review by tushargosavi — COMMENTED** [review]
Please address the provided comment. The reset looks good.

**Comment by tushargosavi** [comment]
Looks good, can you squash all the commits and add JIRA id in the final commit message as per contributing guidelines (http://apex.apache.org/contributing.html). I will merge it after that

**Comment by florianschmidt1994** [comment]
Jenkins, test me again!

**Comment by sanjaypujare** [comment]
I  added the following comment to the JIRA and it will be nice to address it in this PR:

"Can we extend this to all the commands that accept the app-id? I see the following:

connect app-id
get-app-info app-id"

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (sanjaypujare) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:2193]: All the callers throw the identical exception when the returned app is null. Can we just throw that exception here and the caller can always assume app is non-null?

Thinking further, the current behavior - even with my proposed suggestion - fails the whole command if a single app-id/app-name is not found. Do we want to keep that behavior or modify it to only ignore the invalid app-id/app-name

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (sanjaypujare) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Pls see my comment above about failing the command for the whole list when a single app is not found.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (sanjaypujare) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: => "Shutdown app or apps"

**Review by sanjaypujare — CHANGES_REQUESTED** [review]
I suggested extending this to other commands. Let me know what you think

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (florianschmidt1994) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:2193]: The behavior seems to be inconsistent between different CL tools. I just checked  `docker start valid_container_name invalid_container_name`: this would start valid_container but not invalid_container. Then again when using `apt-get install valid_package_name invalid_package_name` it just fails completely, installing neither of those packages.

**Comment by florianschmidt1994** [comment]
Thanks for the reviews! ~Adding the support for the other commands will make it more consistent, so I would go with adding that functionality everywhere.~

EDIT: Actually it is tricky to add support for e.g `get-app-info` with using an app-name, since get-app-info also works for apps that are killed / shutdown / ... already and therefore multiple apps with that name can exist. Maybe open another issue / pr for those cases?

Regarding the question whether a command should fail completely or not if one of the apps is not found, I would go with not running the command at all, the reason being that it feels nicer to me in terms of usability, when one mistypes an app name or app id and wants to make a correction to it afterwards. (Instead of having to retype the whole command only for the failed app-names or app-ids). I'd be happy to get some more input on that though!

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (florianschmidt1994) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:2193]: Making `findApplicationReportFromAppNameOrId()` throw a CLIException could 
- on one hand clean up the code so we don't need to rewrite it every time
- on the other hand it would make it inconsistent with getApplication and getApplicationByName, where no exception is thrown. 

Also if we leave the exception out of the method, we would still have the possibility of implementing the logic for "fail all vs. fail only for invalid appName" in the respective commands, which we do not have if we move it into the method

Comment (sanjaypujare) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:2193]: Ok

**Comment by sanjaypujare** [comment]
Okay, good point about multiple app-ids for the same name. Then it is not worth extending only for connect....

**Comment by vrozov** [comment]
@sanjaypujare @florianschmidt1994 Whether other commands can support application name in addition to application id should be considered separatly for each command. I'd prefer not to block this PR and have separate JIRAs and PRs for each command that can take application name as an argument in addition to application id.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Shutdown application(s) by id or name?

**Comment by tushargosavi** [comment]
I agree with @vrozov to have separate Jiras for supporting app-name in other commands.

**Comment by florianschmidt1994** [comment]
Are there any open issues left to do then?

**Review by tushargosavi — APPROVED** [review]
@sanjaypujare can you merge this if you don't have any additional comments.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:2193]: IMO, it should be the best effort attempt. If application id or name is not found, the error needs to be reported, but it is necessary to continue to the next id/name. At the end, command should report all applications that were shutdown.

**Comment by florianschmidt1994** [comment]
@vrozov I have both the best effort and the existing approach at hand. How / Who will it be decided with option we will go with? Do we need more input from others for that?

**Comment by vrozov** [comment]
@florianschmidt1994 The best approach is to bring the discussion to dev@apex. Once dev community comes to an agreement, summarize the discussion in JIRA.

**Comment by florianschmidt1994** [comment]
Thanks for all your input! As discussed on the mailing list we will go with the "best effort" approach. The "latest" commit (thanks rebase :D ) contains the discussed functionality as well as an additional test to ensure proper behavior.
@sanjaypujare could you have a look at this again?

**Comment by florianschmidt1994** [comment]
Run tests again

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (sanjaypujare) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:2210]: Do we want to check for duplicates?

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (sanjaypujare) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:2371]: Why do we throw exception and stop processing the kill command as opposed to what we decided to do for the shutdown command e.g. continue processing the remaining arguments? In that case we can also achieve some code reuse between shutdown and kill commands - building the app report list etc can be common code.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (sanjaypujare) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:2374]: Same as above - if an exception is thrown the processing stops.

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliShutdownCommandTest.java
Comment (sanjaypujare) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliShutdownCommandTest.java:127]: Wouldn't it be a good idea to flush the current System.out and System.err (or close them) to ensure you get complete/consistent output in your stdOut and stdErr?

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliShutdownCommandTest.java
Comment (sanjaypujare) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliShutdownCommandTest.java:61]: How about adding a test for the kill command as well? You can add it here and rename the class .

**Review by sanjaypujare — CHANGES_REQUESTED** [review]
Pls check the comments

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliShutdownCommandTest.java
Comment (florianschmidt1994) [reply engine/src/test/java/com/datatorrent/stram/cli/ApexCliShutdownCommandTest.java:127]: New squashed version uses autoflush on PrintStream now

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (florianschmidt1994) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:2210]: I'd vote for going with the way it currently behaves, which is shutting down an app if it is found, and if it is specified as argument twice: printing an error message on the second time, because the app is already shut down.

**Comment by florianschmidt1994** [comment]
The current version now implements the best effort approach as discussed on the mailing list. Unfortunately this is now inconsistent to how the kill-app command behaves, being that _kill-app_ stops when an error occurs during killing of an application.

I think there are two ways to go from here now:

1. Merge this PR (when approved), introducing inconsistent behavior for some time and fixing that inconsistent behavior for all other commands in a new JIRA / PR
2. Extend this PR & JIRA Issue to introduce consistent behavior for all commands that support multiple appNames / appIds as arguments

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliShutdownCommandTest.java
Comment (florianschmidt1994) [reply engine/src/test/java/com/datatorrent/stram/cli/ApexCliShutdownCommandTest.java:61]: Added a comment below on how to go from here

**Comment by sanjaypujare** [comment]
@florianschmidt1994  after thinking about this a bit, we should address the kill command inconsistency in a separate JIRA and merge this PR as soon as comments are addressed. Will finish my review....

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (sanjaypujare) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:2210]: ok

**Review by sanjaypujare — APPROVED** [review]
LGTM

**Comment by florianschmidt1994** [comment]
Added a minor change to fix a bug where the cli would wrongfully disconnect from currentApp

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java
Comment (sanjaypujare) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:2245]: In doing this you are assuming that shutdown succeeded or will succeed. If shutdown doesn't succeed why set currentApp to null?

**Comment by pramod** [comment]
Hey guys,

I am @pramod on github but not the Pramod in your team. :-) Could you
please remove me from this thread?

Thanks,
Pramod

**Comment by florianschmidt1994** [comment]
My apologies, sorry about that! Unfortunately I think we can't do that, but you should be able to unsubscribe at the top right of the page, in the section notification. Also your email show have a link called "mute this thread", which should do the same.

**Comment by vrozov** [comment]
@florianschmidt1994 Please rebase

**Comment by florianschmidt1994** [comment]
- [X] rebase @vrozov

**Pull request merged**
